### PR TITLE
Updating the error message for unbound variables. Fixing issue #684.

### PR DESCRIPTION
**Pull request by radumereuta**
@dwightguth @grosu please review.
Now the error message will say: 

```
[Error] Compiler: Unbounded variable X:KList should start with ? or ! to make
it fresh.
    File: D:\work\k\samples\test\test.k
    Location: (6,17,6,18)
    Compilation Phase: class org.kframework.compile.checks.CheckVariables
```

### CODE REVIEW

File: src/javasources/KTool/src/org/kframework/compile/checks/CheckVariables.java [hunk 1]
```diff
@@ -121,19 +122,16 @@ public Void visit(Sentence node, Void _) {
...
-                            "Unbounded Variable " + v.toString() + ".",
-                            getName(), v.getFilename(), v.getLocation()));
-                }
+                if (!options.backend.java())
+                    exType = KException.ExceptionType.ERROR;
+                else
+                    exType = KException.ExceptionType.WARNING;
+                GlobalSettings.kem.register(new KException(exType,
+                        KException.KExceptionGroup.COMPILER,
+                        "Unbounded variable " + v.toString() + "should start with ? or ! to make it fresh.",
```
Comment (grosu) [inline src/javasources/KTool/src/org/kframework/compile/checks/CheckVariables.java]: I would get rid of "to make it fresh".  Other than that, it is exactly what we wanted here.  Thanks, Radu.

**Comment by dwightguth** [comment]
Okay making the behavior change based on the backend here makes no sense.

**Comment by dwightguth** [comment]
Jenkins: test this please

**Comment by grosu** [comment]
I don't know how the implementation is, but this is a general check that needs to be done, not a backend-specific one.  So we should have this check before any backend.

**Comment by traiansf** [comment]
Isn't it possible for the unbounded variable on the right to be universally
quantified as it sometimes happens in logic programming?

2014-06-27 21:32 GMT+03:00 Grigore Rosu notifications@github.com:

**Comment by grosu** [comment]
I don't understand your question, Traian.  Can you be more explicit?  I mean, what does this have to do with the backend?

**Comment by traiansf** [comment]
The error message is basically saying that an unbounded variable on the
right hand side must be existentially quantified (as both ? and ! are).
I was just asking whether it is unconceivable to have unbounded variables
which are universally quantified.

For example, in the Prolog rule
relStar(X,Z) :- rel(X,Y), relStar(Y, Z).
Y is universally quentified, and is also unbounded.  I was asking whether
something like this cannot happen in a K definition.

2014-06-28 16:23 GMT+03:00 Grigore Rosu notifications@github.com:

**Comment by grosu** [comment]
1) In reachability logic (RL), formulae of the form Phi => forall X . Phi' are not that meaningful (in terms of logic programming (LP), the entire RL rule correspons to the conclusion of the LR clause).  What are the transitions that these specify?  There is likely no gamma' that satisfies forall X . Phi'.

2) We want to eventually support the entire RL in K.  When that happens, we will allow qunatifiers in LHS and RHS terms.  For example:

  rule Phi[Y,Z] /\ (forall X . X > Y implies X > Z) => Phi'[Y,Z,?W] /\ (forall X . X > ?W)

So we will have a general solution then.

Grigore

**Comment by radumereuta** [comment]
@andreistefanescu  I removed the check for the java backend just like you suggested. Now I'm always throwing an error message.
@grosu I've also updated the message like you suggested.

**Comment by dwightguth** [comment]
Looks like you need to fix a Java backend test which violates this rule.

**Comment by radumereuta** [comment]
I'm not sure how. @andreiarusoaie @yilongli can you have a look?

**Comment by andreistefanescu** [comment]
Adding ? in from of the variable name in the failing test should do it.

**Pull request merged**